### PR TITLE
Redirect to login page or root if anonymous

### LINKED ISSUE #15315: Redirection to login portal broken when authentication cookie expired
Since anomymous access is disabled on build service on IBS, 

Firefox is complaining every morning on pages already loaded in tabs ( https://build.suse.de/staging_workflows/SUSE:ALP:Source:Standard:1.0 or https://build.suse.de/project/show/SUSE:ALP:Source:Standard:Core:1.0:Build for instance) with the following message:

"can't redirect properly, problem might be disabled or refusing cookies"

and doesn't redirect properly to login page.

I need to open https://build.suse.de/ manually and login from there.

### PR DESCRIPTION
**Pull request by hennevogel**
Everything but root_path requires auth if Configuration.anonymous is true. No need to send people to :back if it will not work. Send them to the correct login page in proxy mode or directly to root_path instead.
    
Fixes #15315

### CODE REVIEW

**Comment by adrianschroeter** [comment]
The request is to return to the page. Going back to main page is the same as closing and reopening the tab.

**Comment by hennevogel** [comment]
> Going back to main page is the same as closing and reopening the tab.

Which is exactly what you have to do now manually....

**Comment by bmwiedemann** [comment]
The whole UX is still improvable, because currently, 
* I follow a link to https://build.suse.de/request/show/314640
* get redirected to /
* log in
* get sent to / again and
* now need to find the original link again, because it is not in the browser history.

One idea of improvement would be to allow a `osc login` to re-use the ssh-key auth for the browser.
Another idea would be to keep the original requested URL and send the user there after login.

**Comment by hennevogel** [comment]
Depends on #15321

**Comment by eliroca** [comment]
> @bmwiedemann:  Another idea would be to keep the original requested URL and send the user there after login.

I've been testing, and found a working patch for this, as mentioned in #15315 :
```
CONFIG['proxy_auth_login_page'] + "?ReturnTo=" + URI.encode_www_form_component(request.fullpath)
```
Then we can work with that `ReturnTo` to redirect the user there.

File: src/api/app/controllers/webui/webui_controller.rb [hunk 1]
```diff
@@ -172,14 +172,20 @@ def require_admin
...
-      end
-    end
+    return if User.session.present?
+    return if ::Configuration.anonymous
+
+    login_page = case CONFIG['proxy_auth_mode']
+                 when :mellon
+                   add_return_to_parameter_to_query(url: CONFIG['proxy_auth_login_page'], paramter_name: 'ReturnTo')
+                 when :ichain
+                   add_return_to_parameter_to_query(url: CONFIG['proxy_auth_login_page'], paramter_name: 'url')
```
Comment (saraycp) [inline src/api/app/controllers/webui/webui_controller.rb:182]:
```suggestion
                   add_return_to_parameter_to_query(url: CONFIG['proxy_auth_login_page'], parameter_name: 'ReturnTo')
                 when :ichain
                   add_return_to_parameter_to_query(url: CONFIG['proxy_auth_login_page'], parameter_name: 'url')
```

File: src/api/app/controllers/webui/webui_controller.rb [hunk 2]
```diff
@@ -240,4 +246,15 @@ def set_influxdb_data
       interface: :webui
     }
   end
+
+  def add_return_to_parameter_to_query(url:, paramter_name:)
+    uri = URI(url)
+    return_to = {}
+    return_to[paramter_name] = request.fullpath
```
Comment (saraycp) [inline src/api/app/controllers/webui/webui_controller.rb:253]:
```suggestion
  def add_return_to_parameter_to_query(url:, parameter_name:)
    uri = URI(url)
    return_to = {}
    return_to[parameter_name] = request.fullpath
```

**Comment by eduardoj** [comment]
There is a typo in the description of the second commit: `Configurtion.anonymous`. It should be `Configuration.anonymous`.

**Review by eduardoj — CHANGES_REQUESTED** [review]
Only a small typo in the description commit. See comment below...

**Review by saraycp — APPROVED** [review]
I did some debugging and ended up with a URL like `http://idp.com?ReturnTo=%2Fproject%2Fshow%2Fhome%3AAdmin` when `Configuration.anonymous` is false and `:mellon` is set.

LGTM

**Comment by hennevogel** [comment]
@eduardoj pushed the typo fix

**Review by eduardoj — APPROVED** [review]

**Pull request merged**